vck-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oauth2/OAuth2Client.kt

```diff
@@ -0,0 +1,174 @@
+package at.asitplus.wallet.lib.oauth2
+
+import at.asitplus.openid.*
+import at.asitplus.openid.OpenIdConstants.CODE_CHALLENGE_METHOD_SHA256
+import at.asitplus.openid.OpenIdConstants.GRANT_TYPE_CODE
+import at.asitplus.signum.indispensable.io.Base64UrlStrict
+import at.asitplus.wallet.lib.iso.sha256
+import at.asitplus.wallet.lib.jws.JwsService
+import at.asitplus.wallet.lib.oidvci.*
+import io.matthewnelson.encoding.core.Encoder.Companion.encodeToString
+import kotlin.random.Random
+
+/**
+ * Simple OAuth 2.0 client to authorize the client against an OAuth 2.0 Authorization Server and request tokens.
+ *
+ * Can be used in OID4VCI flows, e.g. [WalletService].
+ */
+class OAuth2Client(
+    /**
+     * Used to create [AuthenticationRequestParameters], [TokenRequestParameters] and [CredentialRequestProof],
+     * typically a URI.
+     */
+    private val clientId: String = "https://wallet.a-sit.at/app",
+    /**
+     * Used to create [AuthenticationRequestParameters] and [TokenRequestParameters].
+     */
+    private val redirectUrl: String = "$clientId/callback",
+    /**
+     * Used to store the code, associated to the state, to first send [AuthenticationRequestParameters.codeChallenge],
+     * and then [TokenRequestParameters.codeVerifier], see [RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636).
+     */
+    private val stateToCodeStore: MapStore<String, String> = DefaultMapStore(),
+) {
+
+    /**
+     * Send the result as parameters (either POST or GET) to the server at `/authorize` (or more specific
+     * [OAuth2AuthorizationServerMetadata.authorizationEndpoint]).
+     *
+     * Sample ktor code:
+     * ```
+     * val authnRequest = client.createAuthRequest(...)
+     * val authnResponse = httpClient.get(issuerMetadata.authorizationEndpointUrl!!) {
+     *     url {
+     *         authnRequest.encodeToParameters().forEach { parameters.append(it.key, it.value) }
+     *     }
+     * }
+     * val authn = AuthenticationResponseParameters.deserialize(authnResponse.bodyAsText()).getOrThrow()
+     * ```
+     *
+     * @param state to keep internal state in further requests
+     * @param scope in OID4VCI flows the value `scope` from [IssuerMetadata.supportedCredentialConfigurations]
+     * @param authorizationDetails from RFC 9396 OAuth 2.0 Rich Authorization Requests
+     * @param resource from RFC 8707 Resource Indicators for OAuth 2.0, in OID4VCI flows the value
+     * of [IssuerMetadata.credentialIssuer]
+     */
```

Comment:
the code sample is very nice, but don't we already have some ktor dependency in this project, so we could just add the `authnRequest.encodeToParameters().forEach { parameters.append(it.key, it.value) }` as a ready-made extension function in this PR? This obviously affects the other comments too

Reply:
nvm. there is no such dependency. mixed up projects